### DISCUSSION TITLE
Set different speeds and directions to each Crazyflie in a swarm

### DISCUSSION THREAD

**Discussion by Pelochus** [bitcraze/discussions#1393] [Q&A]
So I have something like this:

```python
with MotionCommander(scf, default_height=DEFAULT_HEIGHT) as mc:
    while True:
                # Change direction if passed the limit, also change direction of nominal speed
                if (my_position[0] > SEGMENT_LIMIT) and (my_nominal_speed[scf._link_uri][0] > 0):
                    my_nominal_speed[scf._link_uri][0] = -NOMINAL_SPEED[0]
                elif (my_position[0] < -SEGMENT_LIMIT) and (my_nominal_speed[scf._link_uri][0] < 0):
                    my_nominal_speed[scf._link_uri][0] = NOMINAL_SPEED[0]
    
                # Get nominal (desired standard) speeds in x and y
                vnx = my_nominal_speed[scf._link_uri][0]
    
                print("I'm: ", scf._link_uri, " with X: ", my_position[0], " and with speed: ", vnx)
    
                mc.start_linear_motion(vnx, 0, 0)
                time.sleep(FREQ)
```

This is called in main with 'swarm.parallel()'. I have 2 Crazyflies 2.1 with the Flow Deck V2. This function just tries to have a swarm of Crazyflies going forward and backward from -1 meter to 1 meter, so, when it starts it goes like this:

- Drone starts at 0, goes 1m forward
- When it passes 1m, it goes to -1m (2m backwards)
- It goes to 1m again (2m forward)

My idea is to have the second drone to do the same with a certain delay (say, it starts 3 seconds later)
My problem is that whenever I get into the first if/elif, the direction of both drones changes, even though, as you can see in the code, I try to command different speeds with `my_nominal_speed` which is a dict by URI, composed of XY arrays. 

In other words, this happens:
- Drone 1 gets to 1m before drone 2 and changes direction
- Drone 2 is around 0.5m so it shouldn't change direction
- In reality both drones changes direction
- When they both get to 1 or -1m around the same time, they start to "dance" (going forward and backward repeatedly with the frequency I've set)

And what I want to do is:
- Drone 1 gets to 1m, goes backwards
- Drone 2 still needs to get to 1m, still goes forwards
- When drone 2 gets to 1m, it now changes direction.

Any idea why this is happening? Should I use any other function for commanding movement or speed?

**Comment by gemenerik** [comment]
It doesn't look like `my_position` is actually URI dependent here.

It will be easier for me to debug if you can share your entire Python script. Also check out [this example](https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/swarm/swarmSequence.py) where all Crazyflies in a swarm have their own flight sequence.

**Reply by Pelochus** [reply]
```python
import logging
import time
import threading

import cflib.crtp
from cflib.crazyflie.swarm import CachedCfFactory
from cflib.crazyflie.swarm import Swarm
from cflib.positioning.motion_commander import MotionCommander

uris = {
    'radio://0/80/2M/E7E7E7E701',
    'radio://0/80/2M/E7E7E7E702',
}

# Change which Crazyflie should wait
waiting_uri = 'radio://0/80/2M/E7E7E7E702'

# Only way to access positions easily is making this global (seriously)
all_positions = {}

NOMINAL_SPEED = [0.2, 0] # x, y respectively
DEFAULT_HEIGHT = 0.5 # In meters
SEGMENT_LIMIT = 1
FREQ = 0.2 # How many times we update the speed

# deck_attached_event = Event()
logging.basicConfig(level=logging.ERROR)

# Calculate the speed
# NOTE: For now, this only works for only two drones!
def kuramoto(my_position, other_positions):
    # First of all, adjust parameters as needed
    k = 0.12
    desired_offset = 0

    delta = 0
    for position in other_positions:
        delta += position[0] - my_position[0] - desired_offset
    
    return k * delta

def get_xy_from(radio, positions):
    return (positions[radio].x, positions[radio].y)
    
def coordinated_segment(scf):
    with MotionCommander(scf, default_height=DEFAULT_HEIGHT) as mc:
        my_nominal_speed = {}
        my_nominal_speed[scf._link_uri] = NOMINAL_SPEED

        # To make demos make sense, wait so that they do not start coordinated
        if scf._link_uri == waiting_uri:
            time.sleep(3)

        while True:
            # First we calculate positions
            my_position = get_xy_from(scf._link_uri, all_positions)

            # Made it an array so it can scale in the future
            other_positions = []
            for uri in uris:
                if uri != scf._link_uri:
                    other_positions.append(get_xy_from(uri, all_positions))

            # Change direction if passed the limit, also change direction of nominal speed
            if (my_position[0] > SEGMENT_LIMIT) and (my_nominal_speed[scf._link_uri][0] > 0):
                my_nominal_speed[scf._link_uri][0] = -NOMINAL_SPEED[0]
                print("I'm: ", scf._link_uri, " inside the if")
            elif (my_position[0] < -SEGMENT_LIMIT) and (my_nominal_speed[scf._link_uri][0] < 0):
                my_nominal_speed[scf._link_uri][0] = NOMINAL_SPEED[0]
                print("I'm: ", scf._link_uri, " inside the elif")

            speed_x = my_nominal_speed[scf._link_uri][0]
            speed_y = my_nominal_speed[scf._link_uri][1]
            
            # Add Kuramoto extra speed
            speed_x += kuramoto(my_position, other_positions)

            print("I'm: ", scf._link_uri, " with X: ", my_position[0], " and with speed: ", speed_x)

            mc.start_linear_motion(speed_x, speed_y, 0)
            time.sleep(FREQ)

def recover_positions():
    # Recover positions every FREQ seconds
    global all_positions
    while True:
        all_positions = swarm.get_estimated_positions()
        time.sleep(FREQ / 2) # Update 2 times faster than the speed, just in case

if __name__ == '__main__':
    cflib.crtp.init_drivers()
    factory = CachedCfFactory(rw_cache='./cache')

    with Swarm(uris, factory=factory) as swarm:
        print('Connected to Crazyflies!')

        swarm.reset_estimators()
        print("\nEstimators reset! These are the current positions\n")
        print(swarm.get_estimated_positions())

        # Run threads
        recover_positions_thread = threading.Thread(target=recover_positions)
        
        recover_positions_thread.start()
        time.sleep(0.01)
        swarm.parallel(coordinated_segment)
        
        recover_positions_thread.join()
```

**Reply by Pelochus** [reply]
From what I've debugged, I think that is not the problem. Seems like every time I print each drone commanded speed it is different (and visually speaking, it is too). However when the sign changes, it changes for both at the same time. 

I'll take a look into that example, at first I ignored it but it may be useful.

Thanks!

**Reply by gemenerik** [reply]
I played around with it a little bit, the following makes the delay between the drones work. There must be an error in how you update `my_nominal_speed`.

```python
import logging
import time
import threading

import cflib.crtp
from cflib.crazyflie.swarm import CachedCfFactory
from cflib.crazyflie.swarm import Swarm
from cflib.positioning.motion_commander import MotionCommander

uris = {
    'radio://0/88/2M/E7E7E7E7E8',
    'radio://0/72/2M/E7E7E7E7E8',
}
# Change which Crazyflie should wait
waiting_uri = list(uris)[-1]

# Only way to access positions easily is making this global (seriously)
all_positions = {}

NOMINAL_SPEED = [0.1, 0] # x, y respectively
DEFAULT_HEIGHT = 0.5 # In meters
SEGMENT_LIMIT = 0.5
FREQ = 0.5 # How many times we update the speed

# deck_attached_event = Event()
logging.basicConfig(level=logging.ERROR)

# Calculate the speed
# NOTE: For now, this only works for only two drones!
def kuramoto(my_position, other_positions):
    # First of all, adjust parameters as needed
    k = 0.12
    desired_offset = 0

    delta = 0
    for position in other_positions:
        delta += position[0] - my_position[0] - desired_offset
    
    return k * delta

def get_xy_from(radio, positions):
    return (positions[radio].x, positions[radio].y)
    
def coordinated_segment(scf):
    with MotionCommander(scf, default_height=DEFAULT_HEIGHT) as mc:
        speed_x = NOMINAL_SPEED[0]
        speed_y = NOMINAL_SPEED[1]
        # To make demos make sense, wait so that they do not start coordinated
        if scf._link_uri == waiting_uri:
            time.sleep(3)

        while True:
            # First we calculate positions
            my_position = get_xy_from(scf._link_uri, all_positions)

            # Made it an array so it can scale in the future
            other_positions = []
            for uri in uris:
                if uri != scf._link_uri:
                    other_positions.append(get_xy_from(uri, all_positions))

            # Change direction if passed the limit, also change direction of nominal speed
            if (my_position[0] > SEGMENT_LIMIT) and (speed_x > 0):
                speed_x = -NOMINAL_SPEED[0]
                print("I'm: ", scf._link_uri, " inside the if")
            elif (my_position[0] < -SEGMENT_LIMIT) and (speed_x < 0):
                speed_x = NOMINAL_SPEED[0]
                print("I'm: ", scf._link_uri, " inside the elif")
            
            speed_y = NOMINAL_SPEED[1]
            # Add Kuramoto extra speed
            # speed_x += kuramoto(my_position, other_positions)

            print("I'm: ", scf._link_uri, " with X: ", my_position[0], " and with speed: ", speed_x)

            mc.start_linear_motion(speed_x, speed_y, 0)
            time.sleep(FREQ)

def recover_positions():
    # Recover positions every FREQ seconds
    global all_positions
    while True:
        all_positions = swarm.get_estimated_positions()
        time.sleep(FREQ / 2) # Update 2 times faster than the speed, just in case

if __name__ == '__main__':
    cflib.crtp.init_drivers()
    factory = CachedCfFactory(rw_cache='./cache')

    with Swarm(uris, factory=factory) as swarm:
        print('Connected to Crazyflies!')

        swarm.reset_estimators()
        print("\nEstimators reset! These are the current positions\n")
        print(swarm.get_estimated_positions())

        # Run threads
        recover_positions_thread = threading.Thread(target=recover_positions)
        
        recover_positions_thread.start()
        time.sleep(0.01)
        swarm.parallel(coordinated_segment)
        
        recover_positions_thread.join()
```

**Reply by Pelochus** [reply]
Thank you very much! Gonna try it out as soon as I can.

Quite interesting the very few changes made, what is different from `my_nominal_speed` that breaks the delay?

**Reply by gemenerik** [reply]
I'm not quite sure, I suspect it is a problem in the logic, though I did change some frequencies too while playing around. Let me know if you figure out what made the difference.
